_fr/tour/abstract-type-members.md

```diff
@@ -2,11 +2,80 @@
 layout: tour
 title: Abstract Type Members
 partof: scala-tour
-
-num: 21
-
+num: 25
 language: fr
-
 next-page: compound-types
 previous-page: inner-classes
+topics: abstract type members
+prerequisite-knowledge: variance, upper-type-bound
+
+redirect_from: 
+  - "/tutorials/tour/abstract-types.html"
+  - "/tour/abstract-types.html"
+
 ---
+
+Les types abstraits, tels que les traits et les classes abstraites, peuvent avoir des membres de type abstrait.
+Cela signifie que les implémentations concrètes définissent les types réels.
+Voici un exemple :
+
+```scala mdoc
+trait Buffer {
+  type T
+  val element: T
+}
+```
+
+Ici, nous avons défini un `type T` abstrait. Il est utilisé pour décrire le type de `element`. Nous pouvons étendre ce trait dans une classe abstraite, en ajoutant une limite de type supérieure à `T` pour le rendre plus spécifique.
+
+```scala mdoc
+abstract class SeqBuffer extends Buffer {
+  type U
+  type T <: Seq[U]
+  def length = element.length
+}
+```
+
+Remarquez comment nous pouvons utiliser un autre type abstrait `U` dans la spécification d'un type supérieur pour `T`. Cette `classe SeqBuffer` nous permet de stocker uniquement des séquences dans le tampon en indiquant que le type `T` doit être un sous-type de `Seq[U]` pour un nouveau type abstrait `U`.
```

Comment:
```suggestion
Remarquez comment nous pouvons utiliser un autre type abstrait `U` dans la spécification d'une borne supérieure pour `T`. Cette `classe SeqBuffer` nous permet de stocker uniquement des séquences dans le tampon en indiquant que le type `T` doit être un sous-type de `Seq[U]` pour un nouveau type abstrait `U`.
```

Reply:
```suggestion
Remarquez comment nous pouvons utiliser un autre type abstrait `U` dans la spécification d'un type supérieur pour `T`. Cette `class SeqBuffer` nous permet de stocker uniquement des séquences dans le tampon en indiquant que le type `T` doit être un sous-type de `Seq[U]` pour un nouveau type abstrait `U`.
```
or
```suggestion
Remarquez comment nous pouvons utiliser un autre type abstrait `U` dans la spécification d'un type supérieur pour `T`. Cette classe `SeqBuffer` nous permet de stocker uniquement des séquences dans le tampon en indiquant que le type `T` doit être un sous-type de `Seq[U]` pour un nouveau type abstrait `U`.
```
Backquotes indicate program fragments, in which we should not translate keywords.